lib/launchers/openstack10.rb

```diff
@@ -687,6 +687,50 @@ def assign_ip(instance)
       end
     end
 
+    def allocate_floating_ip(network_name)
+      fipsres = self.rest_run(self.os_network_url + "/v2.0/floatingips", "GET", {}, self.os_token)
+      if fipsres[:success]
+        fips = fipsres.dig(:parsed, "floatingips")
+      else
+        raise "Could not get all existing floating ips\n#{fipsres[:response]}"
+      end
+
+      network = floating_ip_networks.find { |n| n["name"] == network_name }
+      unless network
+        raise "could not find network #{network_name} in current tenant."
+      end
+
+      # filter condition is
+      # 1, floating ip is not belong to given network_name and
+      # 2, floating ip is not preserved
+      # 3, floating ip is not associated to instance
+      filtered_fips = fips.find_all { |n|
+        n["floating_network_id"] == network["id"] && !n["description"][/[Pp]reserve/] && !n["fixed_ip_address"]
+      }
+      unless filtered_fips.empty?
+        return filtered_fips.sample
+      else
+        request_url = self.os_network_url + "/v2.0/floatingips"
```

Comment:
indentation from this line down is deeper

Reply:
Other than that looks good.

One thing to consider, I tried to touch on that with my previous comment, is that if there is time between selecting and using the IP, it is possible to have a conflict between two installations.

e.g. 
1. install_1 selects IP
2. install_2 selects same IP
3. both installations fight at some point for the IP

Reply:
Good point, it will, and only one of two install procedure will success. 
So any suggestion to make a lock?

Reply:
the other option would be to always allocate a new one but deallocate daily old IPs.